plugin-server/src/worker/ingestion/person-state.ts

```diff
@@ -543,55 +550,83 @@ export class PersonState {
-                    {
-                        created_at: firstSeen,
-                        properties: excludeProperties
-                            ? mergeInto.properties
-                            : this.updatedPersonProperties(mergeInto.properties),
-                        is_identified: true,
-                    },
-                    client
-                )
+        try {
+            let mergedPerson
+                // Keep the oldest created_at (i.e. the first time we've seen this person)
+            ;[kafkaMessages, mergedPerson] = await this.handleMergeTransaction(
+                mergeInto,
+                otherPerson,
+                olderCreatedAt,
+                properties
+            )
+            await this.db.kafkaProducer.queueMessages(kafkaMessages)
 
-                // :KLUDGE: Avoid unneeded fetches in updateProperties()
-                this.personContainer = this.personContainer.with(person)
+            // :KLUDGE: Avoid unneeded fetches in updateProperties()
+            this.personContainer = this.personContainer.with(mergedPerson)
+        } catch (error) {
+            if (!(error instanceof DatabaseError)) {
+                throw error // Very much not OK, this is some completely unexpected error
+            }
 
-                // Merge the distinct IDs
-                await this.handleTablesDependingOnPersonID(otherPerson, mergeInto, client)
+            failedAttempts++
+            if (failedAttempts >= MAX_FAILED_PERSON_MERGE_ATTEMPTS) {
+                throw error // Very much not OK, failed repeatedly so rethrowing the error
+            }
 
-                const distinctIdMessages = await this.db.moveDistinctIds(otherPerson, mergeInto, client)
+            // Note this is the persons in the original order (mergeIntoDistinctId and otherPersonDistinctId are never changed),
+            // which is important for property overrides on conflicts consistency
+            await this.aliasDeprecated(
+                otherPersonDistinctId,
+                mergeIntoDistinctId,
+                this.teamId,
+                timestamp,
+                shouldIdentifyPerson,
+                false,
+                failedAttempts,
+                excludeProperties
+            )
+        }
+    }
 
-                const deletePersonMessages = await this.db.deletePerson(otherPerson, client)
+    private isMergeAllowed(mergeFrom: Person): boolean {
+        // $create_alias and $identify will not merge a user who's already identified into anyone else
+        return !mergeFrom.is_identified
+    }
 
-                kafkaMessages = [...updatePersonMessages, ...distinctIdMessages, ...deletePersonMessages]
-            } catch (error) {
-                if (!(error instanceof DatabaseError)) {
-                    throw error // Very much not OK, this is some completely unexpected error
-                }
+    private personAgeInMonthsLowCardinality(timestamp: DateTime): number {
```

Comment:
non-blocking

Now we have this as a function, we could test this to give us some confidence that it is a will stay correct.